### PR TITLE
Fix issues parsing non-standard ERC20 responses

### PR DESCRIPTION
**Pull request by FrederikBolding**
**Description**

Fix an issue parsing non-standard ERC20 responses.

- FIXED:
  - Fix an issue parsing non-standard ERC20 responses

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

**Issue**

Resolves #???

### CODE REVIEW

**Comment by FrederikBolding** [comment]
This may need a sanity check 😅

File: src/assets/Standards/ERC20Standard.ts [hunk 1]
```diff
@@ -58,15 +58,38 @@ export class ERC20Standard {
    * @returns Promise resolving to the 'symbol'.
    */
   async getTokenSymbol(address: string): Promise<string> {
-    const contract = this.web3.eth.contract(abiERC20).at(address);
+    const payload = { to: address, data: '0x95d89b41' };
```
Comment (adonesky1) [inline src/assets/Standards/ERC20Standard.ts:61]: we should add a comment that this is the byte signature for `symbol()`

**Comment by adonesky1** [comment]
Wondering if we should be making the `ERC20Standard` class handle non-standard method signatures... or if we should change our use of the `getTokenStandardAndDetails` method so that we are blocking user action when non-standard tokens are used and handle fetching these values (`symbol` and `decimals`) in their own non-standard class?

**Comment by adonesky1** [comment]
If we do implement these changes for instance, we'll return calls to `getTokenStandardAndDetails` against `MKR` with an object indicating that it is an ERC20 token, which, since it doesn't conform to the ERC20 standard it is not.

**Comment by FrederikBolding** [comment]
> If we do implement these changes for instance, we'll return calls to `getTokenStandardAndDetails` against `MKR` with an object indicating that it is an ERC20 token, which, since it doesn't conform to the ERC20 standard it is not.

I don't think there is a reason to distinguish between MKR (and similar "non-standard ERC20s") and "regular" ERC20s for this specific case. Whether the return value of `symbols` is exactly the correct type should not be blocking imo. I think it should be fine to handle these small variations the same as "regular" ERC20s.

However, we should flesh out support so that we don't just block users from interacting with tokens, just because they don't fit a standard.

File: src/assets/AssetsContractController.test.ts [hunk 2]
```diff
@@ -153,10 +154,36 @@ describe('AssetsContractController', () => {
...
+    const decimals = await assetsContract.getERC20TokenDecimals(
       ERC20_DAI_ADDRESS,
     );
-    expect(Number(symbol)).toStrictEqual(18);
+    expect(Number(decimals)).toStrictEqual(18);
+  });
+
+  it('should support standard ERC20 symbols', async () => {
+    assetsContract.configure({ provider: MAINNET_PROVIDER });
+    const details = await assetsContract.getTokenStandardAndDetails(
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/assets/AssetsContractController.test.ts]: I moved the test to `ERC20Standard` and mocked it.

**Comment by adonesky1** [comment]
> I don't think there is a reason to distinguish between MKR (and similar "non-standard ERC20s") and "regular" ERC20s

I'm not sure I agree with this in general, since the point of a standard is to standardize these interfaces and [the EIP in this case](https://eips.ethereum.org/EIPS/eip-20) specifically indicates that the symbol method, if present, should return a string. But I do think we need to give a bit more thought to what this "non-standard" class should contain. Basically I'm thinking it would just query methods with common enough non-standard token function signatures... So until we've discussed something like that more I'm ok with this approach.

File: src/assets/Standards/ERC20Standard.test.ts [hunk 3]
```diff
@@ -71,4 +72,48 @@ describe('ERC20Standard', () => {
...
+
+  it('should support non-standard ERC20 symbols and decimals', async () => {
+    nock('https://mainnet.infura.io:443', { encodedQueryParams: true })
+      .post('/v3/341eacb578dd44a1a049cbc5f6fd4035', {
+        jsonrpc: '2.0',
+        id: 3,
+        method: 'eth_call',
+        params: [
+          {
+            to: '0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2',
```
Comment (adonesky1) [inline src/assets/Standards/ERC20Standard.test.ts:84]: [nit] - I know we don't do this in the other test but we should.

```suggestion
            to: MKR_ADDRESS,
```

**Review by adonesky1 — APPROVED** [review]
Per my comment [here](https://github.com/MetaMask/controllers/pull/830#issuecomment-1130153832), I think we should consider a different approach to this issue, but for now lets resolve this for these "non-standard" tokens. I will also open a PR in the extension to make sure we're not blocking user action when interacting with these non-standard tokens, regardless of whether or not we find a standard.

**Review by adonesky1 — APPROVED** [review]

**Comment by jveer634** [comment]
Hello @FrederikBolding @adonesky1 
We have a similar problem in our project using metamask.

We have a contract which is not an ERC20 token. But the issue is same. The metamask popup is stuck in the loading screen. I have attached the function we are trying to call and the metamask popup and console output of the metamask popup for the reference

And also I am using Metamask version 10.14.6. 

image
image
image

And this is the output of the sentry api

image

**Comment by FrederikBolding** [comment]
@jveer634 This fix is yet to be released. It will be included in `controllers@30.0.0` and will be bumped in the extension afterwards.

**Pull request merged**